components/embedding_based_laion_retrieval/src/main.py

```diff
@@ -17,23 +18,35 @@
 logger = logging.getLogger(__name__)
 
 
-def query_clip_client(embedding_input, client: ClipClient) -> List[str]:
+def query_clip_client(embeddings: pd.Series, client: ClipClient) -> pd.DataFrame:
     """
-    Given a text query and a ClipClient instance, this function retrieves
-    image URLs and their LAION ids related to the query.
+    Asynchronously retrieve image URLs and ids based on prompts in the provided dataframe.
     Args:
-        embedding_input (TODO): TODO
+        embeddings: Series containing embedding
         client (ClipClient): an instance of ClipClient used to query the images
     Returns:
-        urls: a list of strings, each representing a URL of an image related to the query
+        urls: A dataframe with the image urls and the LAION ids as index
     """
-    try:
-        results = client.query(embedding_input=embedding_input.tolist())
-        urls = [i["url"] for i in results]
-    except ConnectionError as e:
-        urls = ["" for _ in range(len(results))]
 
-    return urls
+    async def async_query():
+        with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
```

Comment:
should we make the `max_workers` a component argument?

Reply:
Sounds like a good idea indeed, but we currently don't have a default mechanism for the arguments. I would add that first.